File: server/devices/ubit.py
# Access to micro:bit, currently for Windows
from server.devices.usb import Usb
import time, json, math, requests
from threading import Thread

_VID = 0x0D28
_PID = 0x0204
_SNOOZE = 0.02
_last_msg = ""
_ubit = Usb("ubit", _VID, _PID)

BUTTON_A = 'a'
BUTTON_B = 'b'
FACE_UP = '^'
FACE_DOWN = 'v'
LEFT = '<'
RIGHT = '>'
UP = 'B'
DOWN = 'F'

def handle_message(msg):
    global _last_msg
    # returns true dictionary result if anything to send
    result = False
    data = False
    try:
        data = json.loads(msg)
    except json.JSONDecodeError:
        data = False
    if type(data) is dict:
        # print("json : " + msg)
        result = {}
        roll = data.get("Ro", False)
        if roll:
            result['roll'] = math.degrees(roll)
        pitch = data.get("Pi", False)
        if pitch:
            result['pitch'] = math.degrees(pitch)
    elif msg != '' and _last_msg != msg: # check for encoded string
        _last_msg = msg
        result = {}
        if BUTTON_A in msg:
            result['button_a'] = True
        if BUTTON_B in msg:
            result['button_b'] = True
        if '0' in msg :
            result['pin_0'] = True
        if '1' in msg :
            result['pin_1'] = True
        if '2' in msg :
            result['pin_2'] = True
        orientation =''
        if FACE_UP in msg :
            orientation = 'up'
        elif FACE_DOWN in msg :
            orientation = 'down'
        elif LEFT in msg :
            orientation = 'left'
        elif RIGHT in msg :
            orientation = 'right'
        elif UP in msg :
            orientation = 'backward'
        elif DOWN in msg :
            orientation = 'forward'
        result['orientation'] = orientation
    return result

def check_message():
    while True:
        line = _ubit.get_line()
        # print(line)
        if line:
            msg = str(line,"utf-8")
            if msg != "":
                _out = handle_message(msg)
                if _out:
                    # print(line.rstrip(), ">>", _out)
                    resp = requests.post('http://127.0.0.1/device/ubit', json = _out)
                    # print(resp.status_code)
        else:
            time.sleep(_SNOOZE)

def run():
    Thread(target=check_message).start()

File: internal/server/devices/usb/ubit/ubit.go
//go:build local || full

package ubit

import (
	"encoding/json"
	"fmt"
	"net/http"
	"quando/internal/server/devices/usb"
	"quando/internal/server/socket"
	"strings"
	"time"

	"go.bug.st/serial"
)

type displayJSON struct {
	Val string `json:"val"`
}

type iconJSON struct {
	Val json.Number `json:"val"`
}

type servoJSON struct {
	Servo json.Number `json:"servo"`
	Angle json.Number `json:"angle"`
}

type serialJSON struct {
	Roll  json.Number `json:"Ro"`
	Pitch json.Number `json:"Pi"`
}

type postJSON struct {
	ButtonA     *bool  `json:"button_a,omitempty"`
	ButtonB     *bool  `json:"button_b,omitempty"`
	Pin0        *bool  `json:"pin_0,omitempty"`
	Pin1        *bool  `json:"pin_1,omitempty"`
	Pin2        *bool  `json:"pin_2,omitempty"`
	Orientation string `json:"orientation,omitempty"`
}

type postAngleJSON struct {
	Roll  json.Number `json:"roll,omitempty"`
	Pitch json.Number `json:"pitch,omitempty"`
}

var lookupGesture = map[string]string{
	"^": "up",
	"v": "down",
	"<": "left",
	">": "right",
	"B": "backward",
	"F": "forward",
}

var serialMode = serial.Mode{
	BaudRate: 115200,
	DataBits: 8,
	Parity:   serial.NoParity,
	StopBits: serial.OneStopBit,
}

var device = usb.Device{
	VID:        "0D28",
	PID:        "0204",
	SerialMode: serialMode,
	NewLine:    "\n",
}

var pressed = new(bool)

func HandleDisplay(w http.ResponseWriter, req *http.Request) {
	display := displayJSON{}
	err := json.NewDecoder(req.Body).Decode(&display)
	if err != nil {
		fmt.Println("Error parsing request", err)
		return
	} else {
		device.Send("D=" + display.Val)
	}
}

func HandleIcon(w http.ResponseWriter, req *http.Request) {
	icon := iconJSON{}
	err := json.NewDecoder(req.Body).Decode(&icon)
	if err != nil {
		fmt.Println("Error parsing request", err)
		return
	} else {
		device.Send("I=" + string(icon.Val))
	}
}

func HandleServo(w http.ResponseWriter, req *http.Request) {
	servo := servoJSON{}
	err := json.NewDecoder(req.Body).Decode(&servo)
	if err != nil {
		fmt.Println("Error parsing request", err)
		return
	} else {
		device.Send("T=" + string(servo.Servo) + "," + string(servo.Angle))
	}
}

func handleButtonGestureTouchPin(msg string) {
	post_json := postJSON{}
	if strings.Contains(msg, "a") {
		post_json.ButtonA = pressed
	}
	if strings.Contains(msg, "b") {
		post_json.ButtonB = pressed
	}
	if strings.Contains(msg, "0") {
		post_json.Pin0 = pressed
	}
	if strings.Contains(msg, "1") {
		post_json.Pin1 = pressed
	}
	if strings.Contains(msg, "2") {
		post_json.Pin2 = pressed
	}
	orientation := ""
	for k, v := range lookupGesture {
		if strings.Contains(msg, k) {
			orientation = v
			break // can only match one orientation (gesture)
		}
	}
	if orientation != "" {
		post_json.Orientation = orientation
	}
	bout, err := json.Marshal(post_json)
	if err != nil {
		fmt.Println("Error marshalling message", err, ":", msg)
	} else {
		str := string(bout)
		prefix := `{"type":"ubit"`
		if str != "{}" {
			prefix += ","
		}
		str = prefix + str[1:]
		socket.Broadcast(str)
	}
}

func handleMessage(msg string) {
	serial_in := serialJSON{}
	err := json.Unmarshal([]byte(msg), &serial_in)
	if err != nil || msg == "" { // i.e. non JSON - so this is encode gesture/button/pin touch
		handleButtonGestureTouchPin(msg)
	} else {
		sendAngle := postAngleJSON{}
		sendAngle.Roll = serial_in.Roll
		sendAngle.Pitch = serial_in.Pitch
		bout, err := json.Marshal(sendAngle)
		if err != nil {
			fmt.Println("Error marshalling message", err, ":", msg)
		} else {
			str := string(bout)
			if str != "{}" {
				prefix := `{"type":"ubit",`
				str = prefix + str[1:]
				socket.Broadcast(str)
				// fmt.Println("  send ", str)
			}
		}
	}
}

func CheckMessages() {
	for {
		response := device.GetLine()
		if response == "" {
			time.Sleep(100 * time.Millisecond)
		} else {
			handleMessage(response)
		}
	}
}

func init() {
	*pressed = true
}
